### DISCUSSION TITLE
Chakra 3: Type '{ children: Element; }' has no properties in common...

### DISCUSSION THREAD

**Discussion by chrhein** [chakra-ui/chakra-ui#9282] [Q&A]
After upgrading to Chakra 3 (3.2.0) I have several components resulting in type errors when passing `children`. 

One example is the `Select` component. Using the exact example from the documentation like the following code snippet, I can't find any solution where passing `children` to these components will work as intended. 

```typescript
<Select collection={collection}>
  <SelectLabel>Select something</SelectLabel>
  <SelectTrigger>
    <SelectValueText placeholder="Placeholder..." />
  </SelectTrigger>
  <SelectContent>
    {collection.items.map((item: CollectionItem) => (
      <SelectItem key={item.value} item={item}>
        {item.label}
      </SelectItem>
    ))}
  </SelectContent>
</Select>
```

![image](https://github.com/user-attachments/assets/5447739b-ce17-4e3a-a76f-83906b12f5c5)

`Select` is created using the snippet tool as suggested in the Chakra documentation. How do I fix these type errors?

**Comment by chrhein** [comment]
The type errors disappear when using the dot annotations introduced with Chakra 3. 

So this works fine:
```typescript
import { Accordion } from "@chakra-ui/react";

<Accordion.Root>
  <Accordion.Item>
    <Accordion.ItemTrigger>
	  {children}
...
```


After creating a Accordion snippet the `AccordionItemTrigger` looks like this:
```typescript
interface AccordionItemTriggerProps extends Accordion.ItemTriggerProps {
  indicatorPlacement?: "start" | "end";
}

export const AccordionItemTrigger = React.forwardRef<HTMLButtonElement, AccordionItemTriggerProps>(
  function AccordionItemTrigger(props, ref) {
    const { children, indicatorPlacement = "end", ...rest } = props;
    return (
      <Accordion.ItemTrigger {...rest} ref={ref}>
        {indicatorPlacement === "start" && (
          <Accordion.ItemIndicator rotate={{ base: "-90deg", _open: "0deg" }}>
            <LuChevronDown />
          </Accordion.ItemIndicator>
        )}
        <HStack gap="4" flex="1" textAlign="start" width="full">
          {children}
        </HStack>
        {indicatorPlacement === "end" && (
          <Accordion.ItemIndicator>
            <LuChevronDown />
          </Accordion.ItemIndicator>
        )}
      </Accordion.ItemTrigger>
    );
  },
);
```

Using this component generated by the snippet results in the same error shown in the screenshot above. For the components only having trouble with not allowing the `children` prop, a workaround is to wrap `AccordionItemTriggerProps` in the Trigger's `forwardRef` with `React.PropsWithChildren`, making the component look like this:
```typescript
export const AccordionItemTrigger = React.forwardRef<
  HTMLButtonElement,
  React.PropsWithChildren<AccordionItemTriggerProps>
>(function AccordionItemTrigger(props, ref) {
...
```

This workaround fixed the type errors for the example, and this code compiles with no errors:
```typescript
import { AccordionRoot, AccordionItem, AccordionItemTrigger } from "@/components/ui/accordion";


<AccordionRoot>
  <AccordionItem>
    <AccordionItemTrigger>
	  {children}
...
```

The error disappears for the `children` related type errors, but other components such as the `SelectValueText` still complains about `placeholder` being an illegal prop. For all other problems I can't find any solution that removes the errors (without slapping `{/* @ts-ignore */}` above each line). 

I aim to support both the old and new syntax to prevent users from having to deal with numerous breaking changes during the upgrade process, so any help would be appreciated.

**Reply by segunadebayo** [reply]
This is a peculiar TS error I've been trying to track down.

What version of `@types/react` do you have?

**Reply by chrhein** [reply]
Have defined `@types/react": "^18.3.12` in my `package.json`

**Reply by chrhein** [reply]
And the one currently in use is `18.3.12`

**Reply by ahcai** [reply]
+1 getting this issue with a lot of the snippets as well (`Select`, `Popover`, etc). Using `React.PropsWithChildren<...>` doesn't fix it for me.

`"@types/react": "^18.0.0",`

**Reply by chrhein** [reply]
It seems to be a problem with most snippet-generated `<X>Trigger` and `<X>Content` for me. A workaround for now that seems to be working is to manually extend interfaces with `ButtonProps` and `BoxProps` for all components that have these errors, resulting in something like this:
```typescript
interface ColorPickerContentProps extends ChakraColorPicker.ContentProps, BoxProps {
...
}
```

This allows me to use the examples from the Chakra documentation without TypeScript errors, where the styling works as intended as well.

**Reply by chrhein** [reply]
Some interfaces like `DrawerContentProps` extends `HTMLChakraProps<"section", ArkDialog.ContentBaseProps>` making them incompatible with the BoxProps extension in my example above (complaining about hundreds of props that has the same name in both types but not identical in use). I still want to append the standard Chakra-props like `padding` and `bg` etc. to these components. For these cases another workaround seems to do the trick by merging the interfaces using this utility type:
```typescript
// Utility type to merge two interfaces, prioritizing properties from B
type Merge<A, B> = Omit<A, keyof B> & B;
```

Using this new `Merge` type I achieve the expected behavior of the `DrawerContent` by defining its interface like this:
```typescript
interface DrawerContentProps extends Merge<BoxProps, ChakraDrawer.ContentProps> {
  ...
}

**Comment by mathisfo** [comment]
I am also experiencing this issue

**Comment by hirish** [comment]
As per my comment on #9038, and in case it helps people in this thread - on my setup I fixed this by updating my `moduleResolution` in tsconfig to `bundler` (from `node`)

**Reply by jgiunta1** [reply]
This should be marked as the answer

**Reply by jgiunta1** [reply]
Switching to "moduleResolution": "bundler" in tsconfig.json resolved the type error for components not accepting children. The issue was due to TypeScript not properly resolving modern exports fields in ESM packages like @ark-ui/react. The bundler mode handles this correctly and picks up the right type definitions.

**Comment by aakashDev7** [comment]
I am using ui package layering in monorepo so the problem with making it with bundler is working for type issues but when i am importing packages with @repo/ui its not working and relative file changes ../../ and so on

**Comment by jieliu218** [comment]
"moduleResolution": "Bundler" in tsconfig

**Comment by segunadebayo** [comment]
Set `moduleResolution` to `Bundler` in your tsconfig. That is what the install guide requires.
